crates/ruma-events/src/call/member.rs

```diff
@@ -0,0 +1,386 @@
+    /// Creates a new `CallMemberEventContent`.
+    pub fn new(memberships: Vec<Membership>) -> Self {
+        Self { memberships }
+    }
+
+    /// All non expired memberships in this member event
+    ///
+    /// # Arguments
+    ///
+    /// * `origin_server_ts` - optionally the `origin_server_ts` can be passed as a fallback in case
+    ///   the Membership does not contain `created_ts`. (`origin_server_ts` will be ignored if
+    ///   `created_ts` is `Some`)
+    pub fn memberships(&self, origin_server_ts: Option<Timestamp>) -> Vec<&Membership> {
+        self.memberships.iter().filter(|m| !m.is_expired(origin_server_ts)).collect()
+    }
+
+    /// All memberships in this event (including expired ones)
+    pub fn memberships_including_expired(&self) -> Vec<&Membership> {
+        self.memberships.iter().collect()
+    }
+
+    /// Each call member event contains the `origin_server_ts` and `content.create_ts`.
+    /// `content.create_ts` is undefined for the initial event of a session (because the
+    /// `origin_server_ts` is not known on the client).
+    /// In the rust sdk we want to copy over the `origin_server_ts` of the event into the content.
+    /// (This allows to use `MinimalStateEvents` and still be able to determine if a membership is
+    /// expired)
+    pub fn set_created_ts_if_none(&mut self, origin_server_ts: Timestamp) {
+        self.memberships.iter_mut().for_each(|m| {
+            m.created_ts = Some(m.created_ts.unwrap_or(origin_server_ts));
+        });
+    }
+}
+
+/// A membership describes one of the sessions this user currently partakes.
+///
+/// The application defines the type of the session.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct Membership {
+    /// The type of the matrixRTC session the membership belongs to. (e.g. call, spacial,
+    /// document...)
+    #[serde(flatten)]
+    pub application: Application,
+    /// The device id of this membership. (The same user can join with their phone/computer)
+    pub device_id: String,
+    /// The duration in milliseconds relative to the time this membership joined
+    /// (`MIN(content.created_ts, event.origin_server_ts)`) during which the membership is valid.
+    pub expires: u64,
+    /// Contains the `origin_server_ts` of the initial session join.
+    /// If the membership is updated this field will be used to track to
+    /// original `origin_server_ts`
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub created_ts: Option<Timestamp>,
+    /// A list of the foci in use for this membership
+    pub foci_active: Vec<Foci>,
+    /// The id of the membership. This is required to guarantee uniqueness of the event.
+    /// (Sending the same state event twice to synapse makes the HS drop the second one and return
+    /// 200)
+    #[serde(rename = "membershipID")]
+    pub membership_id: String,
+}
+
+impl Membership {
```

Comment:
We need a constructor with the minimal set of required fields here, otherwise it won't be possible to construct this type outside of Ruma.

If there are too many required fields, we need to use an exhaustive immutable `MembershipInit` struct that can be converted to a `Membership` (see [`PusherInit`](https://docs.ruma.io/ruma/api/client/push/struct.PusherInit.html) for example).